File: 백준/Silver/12789. 도키도키 간식드리미/도키도키 간식드리미.py
import sys

n = int(sys.stdin.readline())
tickets = list(map(int, sys.stdin.readline().split()))
stack = []
idx = 1
for ticket in tickets:
    stack.append(ticket)
    while stack and stack[-1] == idx:
        stack.pop()
        idx += 1
    if len(stack) > 1 and stack[-2] < stack[-1]:
        break
print('Nice' if not stack else 'Sad')

File: 백준/Silver/12789. 도키도키 간식드리미/도키도키 간식드리미.cc
#include <iostream>
#include <stack>
using namespace std;

int main() {
	cin.tie(0)->sync_with_stdio(0);

	int n;
	cin >> n;
	stack<int> st;
	int num, idx = 1;
	for (int i = 0; i < n; i++) {
		cin >> num;
		if (idx == num) idx++;
		else st.push(num);
		while (!st.empty() && st.top() == idx) {
			st.pop();
			idx++;
		}
	}
	cout << (st.empty() ? "Nice" : "Sad");
}